service/worker/indexer/esProcessor.go

```diff
@@ -0,0 +1,265 @@
+// THE SOFTWARE.
+
+package indexer
+
+import (
+	"context"
+	"encoding/json"
+	"github.com/olivere/elastic"
+	"github.com/uber-common/bark"
+	"github.com/uber/cadence/.gen/go/indexer"
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/collection"
+	"github.com/uber/cadence/common/logging"
+	"github.com/uber/cadence/common/messaging"
+	"github.com/uber/cadence/common/metrics"
+	"time"
+)
+
+type (
+	// ESProcessor is interface for elastic search bulk processor
+	ESProcessor interface {
+		// Stop processor and clean up
+		Stop()
+		// Add request to bulk, and record kafka message in map with provided key
+		Add(request elastic.BulkableRequest, key string, kafkaMsg messaging.Message)
+	}
+
+	// ElasticBulkProcessor is interface for elastic.BulkProcessor
+	// (elastic package doesn't provide such interface that tests can mock)
+	ElasticBulkProcessor interface {
+		Start(ctx context.Context) error
+		Stop() error
+		Close() error
+		Stats() elastic.BulkProcessorStats
+		Add(request elastic.BulkableRequest)
+		Flush() error
+	}
+
+	// esProcessorImpl implements ESProcessor, it's an agent of elastic.BulkProcessor
+	esProcessorImpl struct {
+		processor     ElasticBulkProcessor
+		mapToKafkaMsg collection.ConcurrentTxMap // used to map ES request to kafka message. key: esDocID+visType
+		config        *Config
+		logger        bark.Logger
+		metricsClient metrics.Client
+	}
+)
+
+var _ ESProcessor = (*esProcessorImpl)(nil)
+
+const (
+	// retry configs for es bulk processor
+	esProcessorInitialRetryInterval = 200 * time.Millisecond
+	esProcessorMaxRetryInterval     = 20 * time.Second
+)
+
+// NewESProcessorAndStart create new ESProcessor and start
+func NewESProcessorAndStart(config *Config, client *elastic.Client, processorName string,
+	logger bark.Logger, metricsClient metrics.Client) (ESProcessor, error) {
+	p := &esProcessorImpl{
+		config: config,
+		logger: logger.WithFields(bark.Fields{
+			logging.TagWorkflowComponent: logging.TagValueIndexerESProcessorComponent,
+		}),
+		metricsClient: metricsClient,
+	}
+
+	processor, err := client.BulkProcessor().
+		Name(processorName).
+		Workers(config.ESProcessorNumOfWorkers()).
+		BulkActions(config.ESProcessorBulkActions()).
+		BulkSize(config.ESProcessorBulkSize()).
+		FlushInterval(config.ESProcessorFlushInterval()).
+		Backoff(elastic.NewExponentialBackoff(esProcessorInitialRetryInterval, esProcessorMaxRetryInterval)).
+		After(p.bulkAfterAction).
+		Do(context.Background())
+	if err != nil {
+		return nil, err
+	}
+
+	p.processor = processor
+	p.mapToKafkaMsg = collection.NewShardedConcurrentTxMap(1024, p.hashFn)
+	return p, nil
+}
+
+func (p *esProcessorImpl) Stop() {
+	p.processor.Stop()
+	p.mapToKafkaMsg = nil
+}
+
+// Add an ES request, and an map item for kafka message
+func (p *esProcessorImpl) Add(request elastic.BulkableRequest, key string, kafkaMsg messaging.Message) {
+	if p.mapToKafkaMsg.Contains(key) {
+		kafkaMsg.Ack() // duplicate message
```

Comment:
This looks dangerous, could this result in ack'ing unprocessed message? Can we also store the offset as part of the key to make sure we don't ack unprocessed message.

Reply:
I am not sure how this can result in acking unprocessed message, but I removed this optimization since there is no correctness impact.